# EDA danych dotyczących pasażerów Tytanica

## O Danych

Dane o pasażerach Titanica

Zbiór danych zawiera informacje o pasażerach RMS Titanic, który zatonął 15 kwietnia 1912 roku po zderzeniu z górą
lodową. Dane obejmują takie atrybuty jak klasa podróży, wiek, płeć, liczba rodzeństwa/małżonków na pokładzie,
liczba rodziców/dzieci na pokładzie, cena biletu oraz miejsce zaokrętowania.

Zbiór zawiera także informację o tym, czy pasażer przeżył katastrofę.

Titanic przewoził ponad 2,200 osób, z czego ponad 1,500 zginęło, co czyni tę katastrofę jedną z najbardziej
tragicznych w historii morskiej.

Kolumny:

* **pclass** - Klasa biletu
* **survived** - Czy pasażer przeżył katastrofę
* **name** - Imię i nazwisko pasażera
* **sex** - Płeć pasażera
* **age** - Wiek pasażera
* **sibsp** - Liczba rodzeństwa/małżonków na pokładzie
* **parch** - Liczba rodziców/dzieci na pokładzie
* **ticket** - Numer biletu
* **fare** - Cena biletu
* **cabin** - Numer kabiny
* **embarked** - Port, w którym pasażer wszedł na pokład (C = Cherbourg, Q = Queenstown, S = Southampton)
* **boat** - Numer łodzi ratunkowej
* **body** - Numer ciała (jeśli pasażer nie przeżył i ciało zostało odnalezione)
* **home.dest** - Miejsce docelowe

## A. Wcztywanie danych i ogólny przegląd danych

In [2]:
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
# WCZYTANIE DANYCH
df = pd.read_csv('26__titanic.csv', sep=",")
df

pclass  survived                                             name  \
0        1.0       1.0                    Allen, Miss. Elisabeth Walton   
1        1.0       1.0                   Allison, Master. Hudson Trevor   
2        1.0       0.0                     Allison, Miss. Helen Loraine   
3        1.0       0.0             Allison, Mr. Hudson Joshua Creighton   
4        1.0       0.0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1305     3.0       0.0                            Zabour, Miss. Thamine   
1306     3.0       0.0                        Zakarian, Mr. Mapriededer   
1307     3.0       0.0                              Zakarian, Mr. Ortin   
1308     3.0       0.0                               Zimmerman, Mr. Leo   
1309     NaN       NaN                                              NaN   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000    0.0    0.0   24160  211.3375       B5        S    2   
1       male   0.9167    1.0    2.0  113781  151.5500  C22 C26        S   11   
2     female   2.0000    1.0    2.0  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000    1.0    2.0  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000    1.0    2.0  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1305  female      NaN    1.0    0.0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000    0.0    0.0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000    0.0    0.0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000    0.0    0.0  315082    7.8750      NaN        S  NaN   
1309     NaN      NaN    NaN    NaN     NaN       NaN      NaN      NaN  NaN   

       body                        home.dest  
0       NaN                     St Louis, MO  
1       NaN  Montreal, PQ / Chesterville, ON  
2       NaN  Montreal, PQ / Chesterville, ON  
3     135.0  Montreal, PQ / Chesterville, ON  
4       NaN  Montreal, PQ / Chesterville, ON  
...     ...                              ...  
1305    NaN                              NaN  
1306  304.0                              NaN  
1307    NaN                              NaN  
1308    NaN                              NaN  
1309    NaN                              NaN  

[1310 rows x 14 columns]

In [4]:
# Przegląd kolumn i typów danych
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1310 entries, 0 to 1309
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   float64
 1   survived   1309 non-null   float64
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   float64
 6   parch      1309 non-null   float64
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(7), object(7)
memory usage: 143.4+ KB


In [ ]:
# Wartości unikatowe w kolumnach
df.nunique()

In [20]:
# Wstępna analiza statystyczna danych numerycznych
df.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881166     0.498854     0.385027   
std       0.837836     0.486055    14.413435     1.041658     0.865560   
min       1.000000     0.000000     0.200000     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

           uni_val         fare        body  
count  1309.000000  1308.000000  121.000000  
mean      2.101604    33.295479  160.809917  
std       1.779832    51.758668   97.696922  
min       1.000000     0.000000    1.000000  
25%       1.000000     7.895800   72.000000  
50%       1.000000    14.454200  155.000000  
75%       3.000000    31.275000  256.000000  
max      11.000000   512.329200  328.000000

In [5]:
# USUNIĘCIE WIERSZA o indeksie 1309
df.drop(1309, inplace=True)

In [6]:
# USTAWIENIE WYŚWIETLANIA wszystkich kolumn
pd.set_option('display.max_columns', None)

### Wnioski
1. Baza danych składa się z 1310 wierszy i 14 kolumn (7 z liczbami zmiennoprzecinkowymi i 7 z danymi tekstowymi)
2. Wiersz o indeksie 1309 nie zawiera żadnych danych, więc został usunięty
3. Wyróżniono 3 klasy biletów, 3 porty, w których pasażerowie wsiedli na statek i 27 różnych numerów łodzi ratunkowych (co dziwne, bo jak sprawdziłem mieli ich tylko 20)
4. Najstarszy pasażer miał 80 lat
5. Najwieksza rodzina miała dziwięcioro dzieci
6. Najdroższy bilet kosztował 512,33

## B. Analiza brakujących wartości

In [ ]:
# Sprawdzenie ile wartości brakuje w każdej kolumnie df
df.isnull().sum()

In [ ]:
# Sprawdzenie grupy pasażerów, u których nie ma podanego wieku
df[df["age"].isnull()].sample(10)

In [ ]:
# Sprawdzenie, czy pasażerowie u których nie ma podanej kabiny należą do klasy 3
df[df["cabin"].isnull()].groupby('pclass').size().reset_index(name="count")

### Wnioski
1. W niektórych kolumnach brakuje pojedynczych danych
2. W kolumnie "age" brakuje około 20% danych, nad czym warto będzie się zastanowić
4. Brakuje większości danych w kolumnie "cabin", co nie jest związane z tym, że pasażerowie tańszych klas mogli nie wykupować kajut
5. Brak danych w kolumnie "boat" i "body" nie dziwi, biorąc pod uwagę specyfikę zdarzenia opisanego tą bazą


## C. Analiza pojedynczych zmiennych i transformacja danych

### 1. Klasy biletów

In [7]:
# ZMIANA TYPU ZMIENNYCH w kolumnie "pclass" na liczby całkowite
df['pclass'] = df['pclass'].astype(int)

In [ ]:
# Sprawdzenie ilu pasażerów miało bilety kolejnych klas
pclass_df = df["pclass"].value_counts().reset_index()
pclass_df.columns=['klasa', 'liczba']
pclass_df.sort_values(by='klasa', inplace=True)
pclass_df

In [ ]:
# Generowanie wykresu pokazującego podział pasażerów w zalezności od klasy biletu
pclass_df.plot(kind='bar', x='klasa', y="liczba", legend=False, title="Klasy biletów pasażerów",rot=0, color=["r", "g","b"]) 
plt.show() 

#### Wnioski
* Pasażerów z biletami pierwszej klasy było 323, drugiej klasy 277 a trzeciej klasy 709
* Pasażerów trzeciej klasy było więcej, niż pierwszej i drugiej razem


### 2. Przeżywalność

In [ ]:
# Sprawdzanie ile osób zginęło a ile przeżyło
survived_df = df["survived"].value_counts().reset_index()
survived_df.columns=['przeżywalność', 'liczba']
survived_df['przeżywalność'].replace({0:"martwi", 1:"żywi"}, inplace=True)
survived_df['procent'] = survived_df['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
survived_df

In [ ]:
dead_sex_df = df.groupby(['sex', 'survived']).size().reset_index(name='liczba')
dead_sex_df = dead_sex_df.rename(columns={'sex' : 'płeć'})
# Dodanie kolumny z wyliczeniem wartości procentowej
dead_sex_df['procent'] = dead_sex_df.groupby('płeć')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
dead_sex_df

In [ ]:
# Generowanie wykresu pokazującego ilu pasażerów przeżyło
survived_df.plot(kind='bar', x='przeżywalność', y="liczba",title="Przeżywalność pasażerów", legend=False, color=["r", "g"], rot=0) 
plt.show() 

#### Wnioski
Katastrofę przeżyło 500 pasażerów, czyli 38,2% a zginęło 809, czyli 61,8%

### 3. Nazwiska pasażerów

In [8]:
# PRZETWARZANIE KOLUMNY "name"
# Utworzenie czterech dodatkowych kolumn, roboczej "n_split' oraz "surname" nazwisko, "title" tytuł, "maiden_n" nazwisko panieńskie/rodowe
df.insert(3, 'n_split',"")
df.insert(4, 'surname',"")
df.insert(5, 'title',"")
df.insert(6, 'maiden_n',"")
# Przekazanie danych z oryginalnej kolumny "name" do kolumny roboczej rozdzielonych wcześniej przecinkiem
df['n_split'] = df['name'].str.split(",")
# Przkazanie z kolumny roboczej "n_split" pierwszej części do kolumny "surname" a drugiej części do kolumny "title"
df['surname'] = df['n_split'].apply(lambda x: x[0] if isinstance(x, list) else None)
df['title'] = df['n_split'].apply(lambda x: x[1] if isinstance(x, list) else None)
# Przekazanie z kolumny "title" do "maiden_n" końca zawarotści komórek od ostatniej spacji i bez ostatniego znaku, pod warunkiem, 
# że napis zakończony jest znakiem ')', ale przed którym nie ma znaku'"' (bo tak zaznaczone są przydomki/aliasy)
df['maiden_n'] = df['title'].apply(lambda x: x[::-1].split(' ', 1)[0][::-1][:-1] if isinstance(x, str) and x.endswith(')') and not x[-2] == '"' else None)
# Usunięcie z kolumny "maiden_n" samych imion (w takich przypadkach dane zaczynały się od "(")
df['maiden_n'] = df['maiden_n'].apply(lambda x: x if not isinstance(x, str) or not x.startswith('(') else None)
# Przypisanie do kolumny "title" jej zawartości do pierwszej kropki, co daje sam "tytuł" pasażera
df['title'] = df['title'].str.split('.', n=1).str[0]
# Usunięcie tymczasowej kolumny "n_split"
df.drop('n_split', axis=1, inplace=True)

In [ ]:
# Sprawdzenie ilu było wszystkich pasażerów posiadających nazwiska
df['surname'].notna().sum()

In [ ]:
# Sprawdzenie ile było unikatowych nazwisk pasażerów
df["surname"].nunique()

In [ ]:
# Sprawdzenie ile razy wśród pasażerów powtarzały się nazwiska i które najczęściej
surname_df = df["surname"].value_counts().reset_index().sort_values(by='surname', ascending=False)
surname_df.columns = ['nazwisko', 'pasażerowie']
surname_df.head(5)

In [ ]:
# Sprawdzenie ilu pasażerów dzieliło swojego nazwiska z innymi pasażerami
sur_count_df = surname_df['pasażerowie'].value_counts().reset_index()
sur_count_df.columns = ['liczba nazwisk', 'liczba pasażerów']
sur_count_df['suma pasażerów'] = sur_count_df['liczba nazwisk'] * sur_count_df['liczba pasażerów']
sur_count_df

In [ ]:
# Podział pasażerów w zależności od tego, czy współdzielili nazwisko z kimś innym
sur_pie_df = sur_count_df.groupby(sur_count_df['liczba nazwisk'].apply(lambda x: '1' if x == 1 else 'więcej niż 1'))['suma pasażerów'].sum().reset_index()
sur_pie_df.columns=['pasażerowie', 'liczba pasażerów']
sur_pie_df['pasażerowie'] = ['"samotni"', '"z rodzinami"']
sur_pie_df

In [ ]:
# Generowanie wykresu kołowego pokazującego procentowy udział pasażerów "samotnych" i "z rodzinami"
sur_pie_df.plot.pie(y='liczba pasażerów', labels=sur_pie_df['pasażerowie'], autopct='%1.1f%%', legend=False)
plt.title('Pasażerowie współdzielący nazwiska z innymi lub nie')
plt.show()             

#### Wnioski
* wszystkich pasażerów było 1309
* pasażerowie mieli 875 różnych nazwisk
* najwięcej pasażerów nosiło nazwiska Andersson i Sage, po 11 osób
* 637 pasażerów (48,7 %) nie dzieliło nazwiska z nikim innym, a pozostali 672 (51,3 %) podróżowali z kimś o takim samym nazwisku (byc może z rodziną)

### 4. "Tytuły" pasażerów

In [ ]:
# Sprawdzenie ilu było wszystkich pasażerów posiadających "tytuły"
df['title'].notna().sum()

In [ ]:
# Sprawdzenie ile było unikatowych 'tytułów" pasażerów
df["title"].nunique()

In [ ]:
# Sprawdzenie ile razy wśród pasażerów powtarzały się "tytuły" i które najczęściej
title_df = df["title"].value_counts().reset_index().sort_values(by='title', ascending=False)
title_df.columns = ['tytuł', 'liczba pasażerów']
title_df

#### Wnioski
* każdy z 1309 pasażerów posiadał jakiś "tytuł"
* najbardziej popularne "tytuły" wśród pasażerów: to Mr - Pan (757), Miss - Panna (260), Mrs - Pani (197) i Master - Mistrz/Pan (61)

### 5. Nazwiska rodowe pasażerów

In [ ]:
# Sprawdzenie ilu pasażerów znamy nazwiska rodowe
df['maiden_n'].notna().sum()

In [ ]:
# Sprawdzenie ilu pasażerów, których znamy nazwiska rodowe, to kobiety
maiden_f = ((df['maiden_n'].notna()) & (df['sex'] == 'female')).sum()
print(maiden_f)
# maiden_m = ((df['maiden_n'].notna()) & (df['sex'] == 'male')).sum()
# print(f"Wśród pasażerów, których znamy nazwiska rodowe było {maiden_f} kobiet i {maiden_m} mężczyzn")

In [ ]:
# Sprawdzenie ilu pasażerów, których znamy nazwiska rodowe, to mężczyźni
maiden_m = ((df['maiden_n'].notna()) & (df['sex'] == 'male')).sum()
print(maiden_m)

In [ ]:
# Sprawdzenie ile było unikatowych nazwisk rodowych pasażerów
df["maiden_n"].nunique()

In [ ]:
# Sprawdzenie ile unikatowych nazwisk rodowych jest współdzielonych z nazwiskami pasażerów
matching_values = df['maiden_n'][df['maiden_n'].isin(df['surname'])].unique()
print(len(matching_values))

#### Wnioski
* 181 pasażerów miało nazwiska rodowe, w tym 178 kobiet i 3 mężczyzn
* unikatowych nazwisk rodowych było 169, z czego 45 wystąpiło wśród nazwisk pasażerów

### 6. Płeć

In [ ]:
# Sprawdzenie ilu było pasażerów jakiej płci
sex_df = df["sex"].value_counts().reset_index()
sex_df.columns=['płeć', 'liczba pasażerów']
sex_df['płeć'].replace({"male":"mężczyźni", "female":"kobiety"}, inplace=True)
sex_df

In [ ]:
# Generowanie wykresu kołowego pokazującego podział pasażerów ze względu na płeć
ax = sex_df.plot.pie(y='liczba pasażerów', labels=sex_df['płeć'], autopct='%1.1f%%', legend=False, colors=["yellow", "r"])
plt.title('Podział pasażerów ze względu na płeć')
ax.set_ylabel('')
plt.show()  

#### Wnioski
* wśród pasażerów było 843 mężczyz i 466 kobiet
* mężczyźni stanowili 64,4% pasażerów a kobiety 35,6%

### 7. Wiek

In [ ]:
# Sprawdzenie liczby pasażerów bez podanego wieku
df["age"].isnull().sum()

In [ ]:
# Sprawdzenie ilu pasażerów ma podany wiek
age_df = df.dropna(subset=["age"])
age_df.shape[0]

In [ ]:
# Analiza rozkładu liczebności poszczególnych grup wiekowych pasażerów
ax = df['age'].plot(kind='hist', bins=8, edgecolor='red')
ax.set_title('Histogram wieku pasażerów')
ax.set_xlabel('Grupy wiekowe')
ax.set_ylabel('Ilość pasażerów')
plt.show()

In [9]:
# ZAOKRĄGLENIE WIEKU PASAŻERÓW pasażerów do jednego miejsca po przecinku
df['age'] = df['age'].apply(lambda x: round(x, 1) if pd.notna(x) else x)

In [10]:
# UTWORZENIE NOWEJ KOLUMNY "age_group" z opisem grupy wiekowej
df['age_group'] = df['age'].apply(lambda x: 'dzieci' if x < 18 else ('dorośli' if x >= 18 else None))
# sprawdzenie ilu było wśród pasażerów dorosłych i dzieci
df.groupby('age_group').size().reset_index()

age_group    0
0   dorośli  892
1    dzieci  154

In [ ]:
# Sprawdzenie wieku najstarszego pasażera
df['age'].max()

#### Wnioski
* znamy wiek 1046 pasażerów a nie znamy 264 pasażerów
* rozkład wieku pasażerów jest normalny ze szczytem w grupie 20-30 lat
* wśród pasażerów, których wiek znamy było 892 dorosłych i 154 dzieci
* najstarszy pasażer miał 80 lat

### 8. Bilety, służba i rodziny

In [ ]:
# Sprawdzenie jak rozkładają się wartości w kolumnie "sibsp" - liczba rodzeństwa/małżonków na pokładzie
sibsp_df = df["sibsp"].value_counts().reset_index()
sibsp_df

In [ ]:
# Sprawdzenie jak rozkładają się wartości w kolumnie "parch" - liczba rodziców/dzieci na pokładzie
parch_df = df["parch"].value_counts().reset_index()
parch_df

In [ ]:
# Sprawdzenie u ilu osób podane zostałe informacje o pokrewieństwie
fam_df = df[(df['sibsp'] > 0) | (df['parch'] > 0)]
# fam_df.shape[0]
fam_df

In [ ]:
# Sprawdzenie ilu pasażerów miało bilety
df['ticket'].notna().sum()

In [ ]:
# Sprawdzenie ile zostało kupionych biletów
df["ticket"].nunique()

In [11]:
# PRZETWARZANIE KOLUMNY ZWIĄZANEJ Z BILETAMI
# Wstawienie kolumny z informacją ile osób korzystało z biletu danego pasażera (włącznie z nim)
df.insert(11, 'uni_val',None)
df['uni_val'] = df['ticket'].apply(lambda x: df['ticket'].tolist().count(x))

In [ ]:
# Sprawdzenie ile było biletów grupowych
print(len(df[df['uni_val'] > 1].groupby('ticket').size()))

In [ ]:
# Sprawdzenie ile osób korzystało z biletów grupowych
print((df['uni_val'] > 1).sum())

#### Wnioski cz. 1
* 1309 pasażerów kupiło łącznie 929 bilety, z czego 216 było grupowych (dla 596 osób)
* 519 pasażerów było ze sobą spokrewnionych

In [12]:
# PRZETWARZANIE KOLUMN ZWIĄZANYCH ZE SŁUŻĄCYMI I RODZINAMI
# Wstawienie kolumny dla służących
df.insert(8, 'maid',None)
# Wstawienie kolumny z nazwiskami rodzin
df.insert(8, 'fam',"")
# Wypełnianie kolumny "fam" nazwiskami rodzin, które wystepują wspólnie na bilecie w formacie "nazwisko_numer biletu" i wykazują spokrewnienie
df['fam'] = df.apply(lambda row: (row['surname']+'_'+row['ticket']) if (row['sibsp'] > 0 or row['parch'] > 0) else "", axis=1)

In [13]:
# WYPEŁNIENIE KOLUMNY "maid"
# Wybranie numerów biletów zakupionych przez rodziny
fam_ticket = df[df['fam'] != '']['ticket']
# Wpisanie do kolumny "maid" wartości "1" osobom, które z dużym prawdopodobieństwem pełniły rolę służących
# arunkami wyboru były: obecność na bilecie klasy pierwszej przypisanym do rodziny, brak jakiegokolwiek pokrewieństwa, brak numeru kabiny lub miejsca docelowego
df.loc[(df['ticket'].isin(fam_ticket)) & (df['sibsp'] == 0) & (df['parch'] == 0) & (df['pclass'] == 1) & ((df['cabin'].isna()) | (df['home.dest'].isna())), 'maid'] = 1

In [ ]:
# Utworzenie df zawierającego tylko dane służących
maid_df = df[df['maid'] == 1]

In [ ]:
# Sprawdzenie wieku służących
maid_df['age'].describe()

In [ ]:
# Sprawdzenie płci służących
maid_df.groupby('sex').size()

In [ ]:
# Sprawdzenie "tytułów" służących
maid_df.groupby('title').size()

#### Wnioski cz. 2
* na pokładzie było 23 służących, znamy wiek 20 z nich i wynosił od 22 lat, do 58 lat
* wśród służących było 5 mężczyzn i 18 niezamężnych kobiet (określenie Dona stosowane było raczej do panien)

In [ ]:
# utworzenie df tylko osób spokrewnionych
fam_df = df[(df['sibsp'] > 0) | (df['parch'] > 0)]
# Utworzenie df zawierającego pasżerów spokrewnionych, którzy kupili pojedyncze bilety
single_fam_df = fam_df[fam_df['uni_val'] == 1].reset_index()

In [ ]:
# Sprawdzenie które nazwiska powtarzają się wśród pasażerów, którzy kupili pojedyncze bilety
duplicates_df = single_fam_df[single_fam_df['surname'].apply(lambda x: single_fam_df['surname'].tolist().count(x) > 1)].reset_index()
duplicates_df.drop('level_0', axis=1, inplace=True)
duplicates_df

In [14]:
# W KOLUMNIE "fam" łączenie członków rodzin, którzy kupili pojedyncze bilety
df['fam'] = df['fam'].replace(
    {'Duran y More_SC/PARIS 2149': 'Duran y More_SC/PARIS 2148',
     'Giles_28133' : 'Giles_28134',
     'Andersson_3101281' : 'Andersson_347091',
     'Braund_3460' : 'Braund_A/5 21171',
     'Elias_2690' : 'Elias_2675',
     'Elias_2695' : 'Elias_2675',
     'Gustafsson_3101276' : 'Gustafsson_3101277',
     'Hagland_65303' : 'Hagland_65304',
     'Ilmakangas_STON/O2. 3101270' : 'Ilmakangas_STON/O2. 3101271',
     'Jussila_4136' : 'Jussila_4137',
     'Kiernan_367227' : 'Kiernan_367229',
     'Kink_315152' : 'Kink_315151',
     'Olsen_C 17368' : 'Olsen_4579',
     'Wiklund_3101267' : 'Wiklund_3101266'})

In [ ]:
# Sprawdzenie które nazwiska nie powtarzają się wśród pasażerów, którzy kupili pojedyncze bilety
non_duplicates_df = single_fam_df[single_fam_df['surname'].apply(lambda x: single_fam_df['surname'].tolist().count(x) == 1)].reset_index()
non_duplicates_df.drop('level_0', axis=1, inplace=True)
non_duplicates_df

In [ ]:
# Poszukiwanie krewnych pani Appleton, z domu Lamson
df[df['maiden_n']=='Lamson']

In [ ]:
# Poszukiwanie krewnych pani Crosby
df[df['surname']=='Crosby']

In [ ]:
# Poszukiwanie krewnych Lady Duff Gordon
df[df['surname']=='Duff Gordon']

In [ ]:
# Poszukiwanie krewnych pana Frauenthal
df[df['surname']=='Frauenthal']

In [ ]:
# Poszukiwanie krewnych panny Frolicher
df[df['surname'].str.contains('Frolicher', case=False, na=False)]

In [ ]:
# Poszukiwanie krewnych panny Helen Monypeny Newsom
df[df['name'].str.contains('Monypeny', case=False, na=False)]

In [ ]:
# Sprawdzeni spokrewnionej pani Beckwith
df[df['surname']=='Beckwith']

In [ ]:
# Poszukiwanie krewnych pana Hocking
df[df['surname']=='Hocking']

In [ ]:
# Sprawdzenie spokrewnionej pani Hocking, z domu Needs
df[df['maiden_n']=='Needs']

In [ ]:
# Poszukiwanie krewnych panny Silven
df[df['surname']=='Silven']

In [ ]:
# Sprawdzenie numerów biletów podobnych do numeru biletu panny Silven odnalazło prawdopodobną rodzinę państwa Lahtinen, chociaż ich wiek budzi wątpliwości
df[df['ticket'].str.contains('25065', case=False, na=False)]

In [ ]:
# Poszukiwanie krewnych pana Ware
df[df['surname']=='Ware']

In [ ]:
# Poszukiwanie krewnych panny Andersen-Jensen
df[df['surname'].str.contains('Jensen', case=False, na=False)]

In [ ]:
# Poszukiwanie krewnych panny Bourke
df[df['surname']=='Bourke']

In [ ]:
# Poszukiwanie krewnych pana Davies
df[df['surname']=='Davies']

In [ ]:
# Poszukiwanie krewnych pana Hansen
df[df['surname']=='Hansen']

In [ ]:
# Poszukiwanie krewnych pana Klasen
df[df['surname']=='Klasen']

In [ ]:
# Poszukiwanie krewnych pana Lindqvist
df[df['name'].str.contains('Lindqvist', case=False, na=False)]

In [ ]:
# Sprawdzenie spokrewnionej pani Hirvonen
df[df['surname']=='Hirvonen']

In [ ]:
# Poszukiwanie krewnych pana Persson
df[df['surname']=='Persson']

In [ ]:
# Dalsze poszukiwanie krewnych pana Persson
df[df['maiden_n']=='Persson']

In [ ]:
# Sprawdzenie rodziny pani Lindell, z domu Persson (raczej nie jest spokrewniona z panem Persson)
df[df['fam'].str.contains('Lindell_349910', case=False, na=False)]

In [ ]:
# Sprawdzenie rodziny pani Lindell, z domu Strom (raczej jest spokrewniona z panem Persson)
df[df['fam'].str.contains('Strom_347054', case=False, na=False)]

In [ ]:
# Poszukiwanie krewnych pana Thomas
df[df['surname']=='Thomas']

In [15]:
# W KOLUMNIE "fam" dołączanie pasażerów, którzy kupili pojedyncze bilety do ich rodzin
df['fam'] = df['fam'].replace(
    {'Appleton_11769': 'Brown_11769',
     'Cornell_11770' : 'Brown_11769',
     'Crosby_112901' : 'Crosby_WE/P 5735',
     'Duff Gordon_11755' : 'Duff Gordon_PC 17485',
     'Frauenthal_17765' : 'Frauenthal_PC 17611',
     'Frolicher_13568' : 'Frolicher-Stehli_13567',
     'Newsom_11752' : 'Beckwith_11751',
     'Hocking_29104' : 'Hocking_29105',
     'Wilkes_363272' : 'Hocking_29105',
     'Silven_250652' : 'Lahtinen_250651',
     'Ware_28666' : 'Ware_CA 31352',
     'Jensen_350048' : 'Andersen-Jensen_350046',
     'Bourke_364848' : 'Bourke_364849',
     'Davies_A/4 48873' : 'Davies_A/4 48871',
     'Hansen_350025' : 'Hansen_350026',
     'Klasen_350404' : 'Klasen_350405',
     'Lindqvist_STON/O 2. 3101285' : 'Hirvonen_3101298',
     'Persson_347083' : 'Strom_347054',
     'Thomas_2621' : 'Thomas_2625',
     'Wiklund_3101267' : 'Wiklund_3101266'
    })

In [ ]:
# Utworzenie df zawierającego bilety grupowe osób spokrewnionych o różnych nazwiskach
fam_df = df[(df['sibsp'] > 0) | (df['parch'] > 0)]
group_ticket_df = fam_df[fam_df.groupby('ticket')['fam'].transform(lambda x: x.nunique()) > 1]
group_ticket_df.sort_values(by='ticket')

In [ ]:
# Poszukiwanie krewnych pani Douglas
df[df['surname']=='Douglas']

In [16]:
# W KOLUMNIE "fam" połączenie pasażerów, którzy kupili grupowe bilety i są spokrewnieni, ale mają różne nazwiska
df['fam'] = df['fam'].replace(
    {'Chibnall_113505' : 'Bowerman_113505',
     'Earnshaw_11767' : 'Potter_11767',
     'Schabert_13236' : 'Mock_13236',
     'Hogeboom_13502' : 'Andrews_13502',
     'Shelley_230433' : 'Parrish_230433',
     'Madill_24160' : 'Robert_24160',
     'Stephenson_36947' : 'Eustis_36947',
     'Nicholls_C.A. 33112' : 'Davies_C.A. 33112',
     'Douglas_PC 17558' : 'Douglas_PC 17761'
    })

In [ ]:
# Sprawdzenie krewnych państwa Vander Planke
df[df['surname']=='Vander Planke']

In [17]:
# W KOLUMNIE "fam" połączenie członków rodziny Vander Planke
df['fam'] = df['fam'].replace(
    {'Vander Planke_345764' : 'Vander Planke_345763'})

In [ ]:
# Sprawdzenie pasażerów, których krewnych nie udało się zidentyfikować
fam_df = df[(df['sibsp'] > 0) | (df['parch'] > 0)]
fam_df[fam_df['fam'].duplicated(keep=False) == False]

In [18]:
# W KOLUMNIE "fam" usunięcie informacji o przynależności do rodzin pasażerów, których krewnych nie udało się zidentyfikować
df['fam'] = df['fam'].replace(
    {'Natsch_PC 17596' : '',
     'Hiltunen_250650' : '',
     'Ahlin_7546' : '',
     'Petterson_347076' : ''
    })

In [19]:
# PRZETWARZANIE KOLUMNY "fam" w celu uzyskania nazwy rodziny w formacie nazwisko i numer
# Utworzenie dodatkowej kolumny roboczej "temp"
df.insert(9, 'temp',"")
# Przeniesienie drugiej części zmiennej z kolumny "fam" do "temp"
df['temp'] = df['fam'].str.split('_').str[1]
# Pozostawienie w kolumnie "fam" samego nazwiska rodziny
df['fam'] = df['fam'].str.split('_').str[0]
# Grupowanie i zamiana stringów na numery
df['temp'] = df.groupby('fam')['temp'].transform(lambda x: pd.Categorical(x).codes + 1)
# Zamiana zmiennych float na int a potem na str w kolumnie "temp"
df['temp'] = df['temp'].apply(lambda x: str(int(x)) if isinstance(x, float) and not pd.isna(x) else str(x))
# Dołączanie "numeru" do nazwiska rodziny, jeżeli było więcej niż jedna rodzina o tym samym nazwisku
df['fam'] = df.apply(lambda row: f"{row['fam']} ({row['temp']})" if row['fam'] != "" else row['fam'], axis=1)
# Usunięcie kolumny tymczasowej "temp"
df.drop('temp', axis=1, inplace=True)

In [ ]:
# Analiza wprowadzonej kolumny "fam"
# Sprawdzenie liczby rodzin wśród pasażerów
df['fam'].nunique()

In [ ]:
# Sprawdzenie liczby osób w najliczniejszych rodzinach
fam_df = df["fam"].value_counts().reset_index().sort_values(by='fam', ascending=False)
fam_df.drop(0, inplace=True)
fam_df[:9]

In [ ]:
# Sprawdzenie liczby rodzin o różnej liczebności członków
fam_df['fam'].value_counts().reset_index()

In [ ]:
# Sprawdzenie liczby rodzin z osobami poniżej 18 r.ż.
df.groupby('fam').apply(lambda g: (g['age'] < 18).any()).sum()

#### Wnioski cz. 3
* na statku były 194 rodziny
* najliczniejsze rodziny, to Państwo Sage (11 osób), Goodwin (8 osób), Andersson i Asplund  (po 7 osób), Rice, Panula, Skoog i Fortune (po 6 osób)
* 116 rodzin było 2-osobowych, 54 3-osobowe, 11 4-osobowych, po cztery rodziny 5-osobowe i 6-osobowe, dwie 7-osobowe i po jednej 8-osobowej i 11-osobowej
* w 72 rodzinach były osoby w wieku poniżej 18 lat

### 9. Ceny biletów

In [ ]:
# Spradzenie ilości unikatowych cen biletów
df['fare'].nunique()

In [ ]:
# Sprawdzenie ceny najdroższego biletu
df['fare'].max()

In [ ]:
# Sprawdzenie ceny najtańszego biletu
df['fare'].min()

In [ ]:
# Sprawdzenie pasażerowó, którzy nie płacili za bilet
df[df['fare'] == 0].reset_index()

In [ ]:
# ZAOKRĄGKENIE CENY BILETÓW do dwóch miejsc po przecinku
df['fare'] = df['fare'].apply(lambda x: round(x, 1) if pd.notna(x) else x)

In [ ]:
# PRZETWARZANIE KOLUMNY "fare" w celu uzyskania opłaty za pasażera
# Wstawienie nowej kolumny "fare_pp"
df.insert(14, 'fare_pp',None)
# Przeliczenie ceny biletu na każdego pasażera w zaokrągleniu do dwóch miejsc
df['fare_pp'] = (df['fare']/df['uni_val']).round(1)

In [ ]:
# Sprawdzenie wyliczonych cen biletów za pasażera
df['fare_pp'].describe()

In [ ]:
# Sprawdzenie pasażerów z najdroższym biletem (zarówno jednostkowo, jak i za osobę)
df[df['fare_pp'] == df['fare_pp'].max()]

In [ ]:
# Utworzenie df z danymi o klasach biletów i cenach za osobę
box_df = df[['pclass', 'fare_pp']]
box_df.columns = ['klasa biletu', 'cena']

In [ ]:
# Generowanie wykresu boxplot pokazującego rozkład cen za osobę biletów wszystkich trzech klas
box_df.boxplot(column='cena', by="klasa biletu", vert=False, grid=True)
plt.suptitle('')
plt.title('Porównanie cen za osobę biletów pierwszej, drugiej i trzeciej klasy')
plt.show()

In [ ]:
# Generowanie wykresu boxplot pokazującego rozkład cen za osobę biletów tylko drugiej i trzeciej klasy
box2_3_df = box_df[(box_df['klasa biletu'] == 2) + (box_df['klasa biletu'] == 3)].boxplot(column='cena', by="klasa biletu", vert=False, grid=True)
plt.suptitle('')
plt.title('Porównanie cen za osobę biletów drugiej i trzeciej klasy')
plt.show()

In [ ]:
# Ogólna analiza statystyczna cen biletów klasy pierwszej
box_df[box_df['klasa biletu'] == 1].describe()

In [ ]:
# Ogólna analiza statystyczna cen biletów klasy drugiej
box_df[box_df['klasa biletu'] == 2].describe()

In [ ]:
# Ogólna analiza statystyczna cen biletów klasy trzeciej
box_df[box_df['klasa biletu'] == 3].describe()

#### Wnioski 
* ceny biletów miały 281 unikatowych wartości
* najdroższy bilet kosztował 512,33 i został zakupiony przez rodzinę Cardeza dla 4 osób (co daje również najwyższą cenę za osobę)
* najtańszy bilet nic nie kosztował i takie bilety (wszystkich trzech klas) otrzymało 17 osób
* cena biletu za osobę wynosiła od 0 do 128, ze średnią 14,75 i medianą 8,05
* rozkład cen za osobę  wykazuje liczne wartości skrajne jedynie w klasie 3
* w klasie pierwszej podróżowały 323 osoby średnio za 33,91 z medianą 29,7
* w klasie drugiej podróżowało 277 osób średnio za 11,41 z medianą 12,35
* w klasie trzeciej podróżowało 708 osób średnio za 7,33 z medianą 7,75
* bilety pierwszej klasy były około 3 razy droższe niż drugiej klasy i ponad 4 razy droższe od trzeciej klasy
* bilety drugiej klasy były o około 50% droższe od trzeciej klasy

### 10. Numery kabiny

In [ ]:
# Sprawdzenie ilu pasażerów ma przypisane kabiny
df['cabin'].notna().sum()

In [ ]:
# Sprawdzenie do jakich klas biletów są przypisane kabiny
df[df['cabin'].notna()].groupby('pclass').size().reset_index()

In [ ]:
# Utworzenie df zawierającego tylko pasażerów z przypisanymi kabinami
cab_df = df[df['cabin'].notna()]
cab_df.sample(10)

In [ ]:
# Sprawdzenie pasażerów z przypisanymi kabinami z biletami trzeciej klasy
cab_df[cab_df['pclass'] == 3]

In [ ]:
# Sprawdzenie pasażerów z przypisanymi kabinami z biletami drugiej klasy
cab_df[cab_df['pclass'] == 2]

#### Wnioski
* jedynie 295 pazsażerów ma przypisane kabiny
* kabiny są przypisane do 256 pasażerów z biletami pierwszej klasy, 23 z drugiej klasy i 16 z trzeciej klasy
* oznaczenie kabiny skałada się z dużej litery alfabetu i dołączonej liczby, pierwsze litery alfabetu i niższe liczby świadczą zwykle o wyższej klasie biletu
* niektórzy pasażerowie z biletami pierwszej klasy mają przypisanych kilka kabin

### 11. Port początkowy

In [ ]:
# Sprawdzenie ilu pasażerów ma przypisany port początkowy
df['embarked'].notna().sum()

In [ ]:
# Sprawdzenie ilu pasażerów wsiadało w których portach
df[df['embarked'].notna()].groupby('embarked').size().reset_index()

In [ ]:
# Sprawdzenie w którym porcie weszło na pokład ilu pasażerów każdej klasy biletów
emb_df= df.groupby(['embarked', 'pclass']).size().reset_index()
emb_df.columns = ['porty', 'klasa biletu', 'liczba']
stack_df = emb_df.pivot(index='porty', columns='klasa biletu', values='liczba')
stack_df

In [ ]:
# Generowanie wykresu pokazującego, w którym porcie weszło na pokład ilu pasażerów każdej klasy biletów
ax = stack_df.plot(kind='bar', stacked=True)
plt.title('Zaokrętowanie pasażerów różnych klas w portach')
plt.ylabel('Liczba pasażerów')
plt.xticks(rotation=0)
ax.set_xticklabels(['Cherbourg','Queenstown','Southampton'])
plt.show()

#### Wnioski
* znamy port zaokrętowania 1307 pasażerów
* najwięcej pasażerów weszło na pokład w Southampton - 914, najmniej w Queenstown - 123, a w Cherbourg - 270
* pasażerowie z biletami pierwszej klasy weszli na pokład głównie w Cherbourg i w Southampton
* w Queenstown zaokrętowali w zasadzie jedynie pasażerowie trzeciej klasy
* w Southampton zaokrętowali głównie pasażerowie drugiej i trzeciej klasy

### 12. Łodzie ratunkowe

In [ ]:
# Sprawdzenie ilu pasażerów ma przypisane łodzie ratunkowe
df['boat'].notna().sum()

In [ ]:
# Spradzenie ilości unikatowych oznaczeń łodzi ratunkowych
df['boat'].nunique()

In [ ]:
# Sprawdzenie jakie są unikatowe oznaczenia łodzi ratunkowych
df['boat'].unique()

In [ ]:
# Sprawdzenie pasażerów, którzy mają przypisana więcej, niż jedną łódź ratunkową
df[df['boat'].str.contains(' ', case=False, na=False)]

In [ ]:
# ZMIANA OZNACZENIA ŁODZI - zostawiam oznaczenie ostatniej łodzi
df['boat'] = df['boat'].apply(lambda x: x.rsplit(' ', 1)[-1] if isinstance(x, str) and ' ' in x else x)

In [ ]:
# Sprawdzenie ilości unikatowych oznaczeń łodzi ratunkowych po wprowadzeniu zmian
df['boat'].nunique()

In [ ]:
# Utworzenie df z pogrupowanymi danymi o łodziach ratunkowych
boat_df = df.groupby('boat').size().reset_index()
# Ogólna analiza statystyczna danych o łodziach ratunkowych
boat_df.describe()

In [ ]:
# Utworzenie df z pogrupowanymi łodziami i klasami biletów pasażerów
boat_class_df = df.groupby(['boat', 'pclass']).size().reset_index()
# Dodanie do pojedynczych numerów łodzi "0" w celu ułatwienia sortowania
boat_class_df['boat'] = boat_class_df['boat'].apply(lambda x: x.zfill(2) if x.isdigit() else x)
# Sortowanie danych według poprawionych oznaczeń łodzi
boat_class_df = boat_class_df.sort_values(by='boat')
# Zmiana nazw kolumn
boat_class_df.columns = ['Łódź', 'klasa biletu', 'liczba']

In [ ]:
# Utworzenie df do wykresu kolumnowego ze stakowaniem
class_stack_df = boat_class_df.pivot(index='Łódź', columns='klasa biletu', values='liczba')
# Wypełnienie pustych miejsc w df wartościami "0"
class_stack_df = class_stack_df.fillna(0)

In [ ]:
# Generowanie wykresu pokazującego 
ax = class_stack_df.plot(kind='bar', stacked=True)
plt.title('Pasażerowie różnych klas w łodziach ratunkowych')
plt.ylabel('Liczba pasażerów')
plt.xticks(rotation=0)
plt.show()

In [ ]:
# Utworzenie df z pogrupowanymi łodziami i płcią  pasażerów
boat_sex_df = df.groupby(['boat', 'sex']).size().reset_index()
# Dodanie do pojedynczych numerów łodzi "0" w celu ułatwienia sortowania
boat_sex_df['boat'] = boat_sex_df['boat'].apply(lambda x: x.zfill(2) if x.isdigit() else x)
# Sortowanie danych według poprawionych oznaczeń łodzi
boat_sex_df = boat_sex_df.sort_values(by='boat')
# Zmiana nazw kolumn
boat_sex_df.columns = ['Łódź', 'Płeć', 'liczba']

In [ ]:
# Utworzenie df do wykresu kolumnowego ze stakowaniem
sex_stack_df = boat_sex_df.pivot(index='Łódź', columns='Płeć', values='liczba')
# Wypełnienie pustych miejsc w df wartościami "0"
sex_stack_df = sex_stack_df.fillna(0)

In [ ]:
# Generowanie wykresu pokazującego ilość pasażerów róznej płci w łodziach ratunkowych
ax = sex_stack_df.plot(kind='bar', stacked=True)
plt.title('Pasażerowie różnej płci w łodziach ratunkowych')
plt.ylabel('Liczba pasażerów')
plt.xticks(rotation=0)
ax.legend(['kobiety','mężczyźni'])
plt.show()

In [ ]:
# Utworzenie df z pogrupowanymi łodziami i podziałem na dzieci i dorosłych
boat_age_df = df.groupby(['boat', 'age_group']).size().reset_index()
# Dodanie do pojedynczych numerów łodzi "0" w celu ułatwienia sortowania
boat_age_df['boat'] = boat_age_df['boat'].apply(lambda x: x.zfill(2) if x.isdigit() else x)
# Sortowanie danych według poprawionych oznaczeń łodzi
boat_age_df = boat_age_df.sort_values(by='boat')
# Zmiana nazw kolumn
boat_age_df.columns = ['Łódź', 'Wiek', 'liczba']

In [ ]:
# Utworzenie df do wykresu kolumnowego ze stakowaniem
age_stack_df = boat_age_df.pivot(index='Łódź', columns='Wiek', values='liczba')
# Wypełnienie pustych miejsc w df wartościami "0"
age_stack_df = age_stack_df.fillna(0)

In [ ]:
# Generowanie wykresu pokazującego ilość dorosłych i dzieci w łodziach ratunkowych
ax = age_stack_df.plot(kind='bar', stacked=True)
plt.title('Dorośli i dzieci w łodziach ratunkowych')
plt.ylabel('Liczba pasażerów')
plt.xticks(rotation=0)
ax.legend(['dorośli','dzieci'])
plt.show()

#### Wnioski
* 486 pasażerów ma przypisane miejsca w łodziach ratunkowych
* łodzi ratunkowych było 20, 16 opisanych kolejnymi numerami od 1 do 16 i 4 oznaczone dużymi litarami od A do D
* w łodziach znajdowało się od 5 do 39 pasażerów, średnio 24,3 pasażera, mediana wynosiła 25
* łodzie o numerach od 1 do 8 zajmowali głównie pasażerowie pierwszej klasy
* wiekszość pasażerów drugiej klasy zajmowało łodzie o numerach od 9 do 14
* pasażerowie trzeciej klasy zajmowali łodzie numer 13, 15 i 16 oraz te oznaczone litarami alfabetu
* w większości łodzi kobiety stanowiły ponad połowę pasażerów
* w łodzi numer 8 były same kobiety a w łodzi B sami mężczyźni
* w wiekszości łodzi znajdowały się dzieci

### 13. Ciała pasażerów

In [ ]:
# Sprawdzenie ile odnaleziono ciał pasażerów
df['body'].notna().sum()

In [ ]:
# Spradzenie czy numery ciał nie powtarzają się
df['body'].nunique() == df['body'].notna().sum()

In [ ]:
# Utworzenie df zawierającego tylko dane pasażerów, których ciała zostały odnalezione
body_df = df[df['body'].notna()]

In [ ]:
# Sprawdzenie ile było dzieci wśród pasażerów, których ciała odnaleziono
(body_df['age'] < 18).sum()

In [ ]:
# Sprawdzenie ile było kobiet wśród pasażerów, których ciała odnaleziono
(body_df['sex'] == "female").sum()

In [ ]:
# Utworzenie df zawierającego tylko dane pasażerów, którzy zginęli
dead_df = df[df['survived'] == 0]

In [ ]:
# Generowanie wykresu kołowego pokazującego ile odnaleziono ciał ofiar
dead_pie_df = dead_df.groupby(df['body'].notna()).size().reset_index()
dead_pie_df.columns=['ciała', 'liczba pasażerów']
dead_pie_df['ciała'].replace({False : "zaginione", True :"odnalezione"}, inplace=True)
dead_pie_df

In [ ]:
# Generowanie wykresu kołowego pokazującego ilość zaginionych i odnalezionych ciał pasażerów
ax = dead_pie_df.plot.pie(y='liczba pasażerów', labels=dead_pie_df['ciała'], autopct='%1.1f%%', legend=False, colors=["r", "g"])
plt.title('Ciała ofiar katastrofy')
ax.set_ylabel('')
plt.show()  

#### Wnioski
* odnaleziono i ponumerowano ciała 121 ofiar katastrofy
* wśród odnalezionych ciał było 8 kobiet i pięcioro dzieci
* odnaleziono ciała 15% wszystkich ofiar, ciała 688 ofiar nie zostały odnalezione

### 14. Miejsca docelowe

In [ ]:
# Sprawdzenie ilu pasażerów miało podane miejsce docelowe
df['home.dest'].notna().sum()

In [ ]:
# Sprawdzenie ile było unikatowych miejsc docelowych
df['home.dest'].nunique()

In [ ]:
# Sprawdzenie częstości wystepowania miejsc docelowych
df['home.dest'].value_counts().reset_index().head()

#### Wnioski
* 745 pasażerów miało podane miejsce docelowe
* unikatowych miejsc docelowych było 369
* najczęstszymi miejscami docelowymi był "New York, NY" - 64 osoby, "London" - 14 osób i "Montreal, PQ" - 10 osób

## D. Analiza relacji między zmiennymi - pytania

### 1. Czy hasło "ratujemy najpierw kobiety i dzieci" zadziałało?

In [ ]:
# Zebranie danych dotyczących płci i przeżywalności
dead_sex_df = df.groupby(['sex', 'survived']).size().reset_index(name='liczba')
dead_sex_df = dead_sex_df.rename(columns={'sex' : 'płeć'})
# Dodanie kolumny z wyliczeniem wartości procentowej
dead_sex_df['procent'] = dead_sex_df.groupby('płeć')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
dead_sex_df

In [ ]:
# Przygotowanie danych dla wykresu pokazującego przeżywalność kobiet i mężczyzn
dead_sex_stack_df = dead_sex_df.pivot(index='płeć', columns='survived', values='liczba')
dead_sex_stack_df

In [ ]:
# Generowanie wykresu pokazującego przeżywalność kobiet i mężczyzn
dead_sex_stack_df = dead_sex_df.pivot(index='płeć', columns='survived', values='liczba')
ax = dead_sex_stack_df.plot(kind='bar', stacked=True, color=['b', 'green'])
plt.title('Przeżywalność pasażerów w zależności od płci')
plt.ylabel('Liczba pasażerów')
plt.xticks(rotation=0)
ax.set_xticklabels(['kobiety', 'mężczyźni'])
ax.legend(['zginęli','przeżyli'])
plt.show()

#### Wnioski cz.1
* katastrofę przeżyło 339 kobiet, co stanowi 72,7% wszystkich kobiet wśród pasażerów
* katastrofę przeżyło jedynie 161 mężczyzn, co stanowi 19,1% wszystkich mężczyzn wśród pasażerów
* biorąc pod uwagę wartości procentowe, katastrofę przeżyło prawie 4 razy więcej kobiet, niż mężczyzn

In [ ]:
# Zebranie danych dotyczących grupy wiekowej i przeżywalności
dead_age_df = df.groupby(['age_group', 'survived']).size().reset_index(name='liczba')
dead_age_df = dead_age_df.rename(columns={'age_group' : 'grupa wiekowa'})
# Dodanie kolumny z wyliczeniem wartości procentowej
dead_age_df['procent'] = dead_age_df.groupby('grupa wiekowa')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
dead_age_df

In [ ]:
# Przygotowanie danych dla wykresu pokazującego przeżywalność dorosłych i dzieci
dead_age_stack_df = dead_age_df.pivot(index='grupa wiekowa', columns='survived', values='liczba')
dead_age_stack_df

In [ ]:
# Generowanie wykresu pokazującego przeżywalność dorosłych i dzieci
dead_age_stack_df = dead_age_df.pivot(index='grupa wiekowa', columns='survived', values='liczba')
ax = dead_age_stack_df.plot(kind='bar', stacked=True, color=['b', 'green'])
plt.title('Przeżywalność pasażerów w zależności od grupy wiekowej')
plt.ylabel('Liczba pasażerów')
plt.xticks(rotation=0)
ax.set_xticklabels(['dorośli', 'dzieci'])
ax.legend(['zginęli','przeżyli'])
plt.show()

#### Wnioski cz.2
* z pasażerów dorosłych, których znamy wiek (892 osób) przeżyło 346, czyli 38,8%
* z dzieci, których znamy wiek (154 osoby) przeżyło 81, czyli 52,6%
* biorąc pod uwagę wartości procentowe, dzieci przeżyło niewiele więcej niż pasażerów dorosłych 

### 2. Czy klasa biletu miała wpływ na przeżywalność?

In [ ]:
# Zebranie danych dotyczących klas biletów i przeżywalności
dead_pclass_df = df.groupby(['pclass', 'survived']).size().reset_index(name='liczba')
dead_pclass_df = dead_pclass_df.rename(columns={'pclass' : 'klasa'})
# Dodanie kolumny z wyliczeniem wartości procentowej
dead_pclass_df['procent'] = dead_pclass_df.groupby('klasa')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
dead_pclass_df

In [ ]:
# Przygotowanie danych dla wykresu pokazującego przeżywalność pasażerów z biletami różnych klas
dead_pclass_pie_df = dead_pclass_df.pivot(index='klasa', columns='survived', values='liczba')
dead_pclass_pie_df

In [ ]:
# Generowanie wykresu pokazującego przeżywalność pasażerów z biletami różnych klas
fig, axes = plt.subplots(1, 3, figsize=(10, 3.5))
fig.suptitle('Przeżywalność pasażerów z biletami różnych klas', fontsize=14)
# Iterowanie przez wiersze df
for i, (index, row) in enumerate(dead_pclass_pie_df.iterrows()):
    axes[i].pie(row, labels=['zginęli','przeżyli'], autopct='%1.1f%%', startangle=180)
    axes[i].set_title(f'klasa biletu: {index}')
    axes[i].axis('equal')
# Dopasowanie wykresów, żeby się nie nachodziły
plt.tight_layout()
plt.show()

#### Wnioski cz.1
* z 323 pasażerów z biletami pierwszej klasy przeżyło 200, czyli 61,9%
* z 277 pasażerów z biletami drugiej klasy przeżyło 119, czyli 43%
* z 709 pasażerów z biletami trzeciej klasy przeżyło 181, czyli 25,5%
* porównując wartości procentowe, pasażerów klasy pierwszej przeżyło o połowę więcej niż klasy drugiej i ponad dwa razy więcj, niż klasy trzeciej

In [ ]:
# Utworzenie df zawierającego dane wyłącznie kobiet
fem_df = df[df['sex'] == 'female']

In [ ]:
# Zebranie danych dotyczących klas biletów i przeżywalności kobiet
fem_dead_pclass_df = fem_df.groupby(['pclass', 'survived']).size().reset_index(name='liczba')
fem_dead_pclass_df = fem_dead_pclass_df.rename(columns={'pclass' : 'klasa'})
# Dodanie kolumny z wyliczeniem wartości procentowej
fem_dead_pclass_df['procent'] = fem_dead_pclass_df.groupby('klasa')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
fem_dead_pclass_df

In [ ]:
# Przygotowanie danych dla wykresu pokazującego przeżywalność kobiet z biletami różnych klas
fem_dead_pclass_pie_df = fem_dead_pclass_df.pivot(index='klasa', columns='survived', values='liczba')
fem_dead_pclass_pie_df

In [ ]:
# Generowanie wykresu pokazującego przeżywalność kobiet z biletami różnych klas
fig, axes = plt.subplots(1, 3, figsize=(10, 3.5))
fig.suptitle('Przeżywalność kobiet z biletami różnych klas', fontsize=14)
# Iterowanie przez wiersze df
for i, (index, row) in enumerate(fem_dead_pclass_pie_df.iterrows()):
    axes[i].pie(row, labels=['zginęły','przeżyły'], autopct='%1.1f%%', startangle=180)
    axes[i].set_title(f'klasa biletu: {index}')
    axes[i].axis('equal')
# Dopasowanie wykresów, żeby się nie nachodziły
plt.tight_layout()
plt.show()

#### Wnioski cz.2
* ze 144 kobiet z biletami pierwszej klasy przeżyło 139, czyli 96,5% 
* ze 106 kobiet z biletami drugiej klasy przeżyły 94, czyli 88,7%
* z 216 kobiet z biletami trzeciej klasy przeżyło 106, czyli 49,1%
* kobiety z biletami klasy pierwszej i drugiej przeżyły prawie wszystkie
* z kobiet z biletami klasy trzeciej przeżyło nieco mniej niż połowa

In [ ]:
# Utworzenie df zawierającego dane wyłącznie dzieci
child_df = df[df['age'] < 18]

In [ ]:
# Zebranie danych dotyczących klas biletów i przeżywalności kobiet
ch_dead_pclass_df = child_df.groupby(['pclass', 'survived']).size().reset_index(name='liczba')
ch_dead_pclass_df = ch_dead_pclass_df.rename(columns={'pclass' : 'klasa'})
# Dodanie kolumny z wyliczeniem wartości procentowej
ch_dead_pclass_df['procent'] = ch_dead_pclass_df.groupby('klasa')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
ch_dead_pclass_df

In [ ]:
# Przygotowanie danych dla wykresu pokazującego przeżywalność dzieci z biletami różnych klas
ch_dead_pclass_pie_df = ch_dead_pclass_df.pivot(index='klasa', columns='survived', values='liczba')
ch_dead_pclass_pie_df

In [ ]:
# Generowanie wykresu pokazującego przeżywalność dzieci z biletami różnych klas
fig, axes = plt.subplots(1, 3, figsize=(10, 3.5))
fig.suptitle('Przeżywalność dzieci z biletami różnych klas', fontsize=14)
# Iterowanie przez wiersze df
for i, (index, row) in enumerate(ch_dead_pclass_pie_df.iterrows()):
    axes[i].pie(row, labels=['zginęły','przeżyły'], autopct='%1.1f%%', startangle=180)
    axes[i].set_title(f'klasa biletu: {index}')
    axes[i].axis('equal')
# Dopasowanie wykresów, żeby się nie nachodziły
plt.tight_layout()
plt.show()

#### Wnioski cz.3
* z 15 dzieci z biletami pierwszej klasy przeżyło 13, czyli 86,7% 
* z 33 dzieci z biletami drugiej klasy przeżyło 29, czyli 87,9%
* ze 106 dzieci z biletami trzeciej klasy przeżyło tylko 39, czyli 36,8%
* dzieci z biletami klasy pierwszej i drugiej przeżyły prawie wszystkie
* z dzieci z biletami klasy trzeciej przeżyło nieco więcej niż 1/3

### 3. Jak przeżywalność zależała od wieku pasażerów?

In [ ]:
# Sprawdzenie procentowej przeżywalności pasażerów w 8 kolejnych grupach wiekowych
age_surv_df = df.groupby(pd.cut(df['age'], bins=8, right=True)).apply(lambda x: x['survived'].value_counts(normalize=True).mul(100)).unstack(fill_value=0)
# Przypisanie nowych indeksów
age_surv_df.index = ['0-10', '10-20', '20-30', '30-40', '40-50', '50-60', '60-70', '70-80']
age_surv_df

In [ ]:
# Generowanie wykresu liniowego pokazującego przeżywalności pasażerów w 8 kolejnych grupach wiekowych
age_surv_df[1].plot(kind='line')
plt.xlabel('grupa wiekowa')
plt.ylabel('procent')
plt.title('Przeżywalności procentowa pasażerów w grupach wiekowych')
plt.show()

In [ ]:
# Sprawdzenie liczebności grup wiekowych
df.groupby(pd.cut(df['age'], bins=8, right=True)).size().reset_index()

#### Wnioski
* najwiekszą przeżywalność miały dzieci w wieku do 10 lat - 58%
* kolejne grupy wiekowe 10-60 lat miały podobną przeżywalność w zakresie 37-48% z niewielkim trendem wzrostowym
* w grupach wiekowych 60-80 lat przeżywalność była bardzo niska 22-33%, jednak w grupie tej znajdowało się mało pasażerów (tylko 33)

### 4. Czy rodziny spotkał taki sam los, co osoby niespokrewnione?

In [ ]:
# Zebranie danych dotyczących obecności rodziny pasażera i przeżywalności
fam_dead_df = df.groupby([df['fam'].apply(lambda x: 'samotni' if x == '' else 'spokrewnieni'), 'survived']).size().reset_index(name='liczba')
fam_dead_df = fam_dead_df.rename(columns={'fam' : 'status'})
# Dodanie kolumny z wyliczeniem wartości procentowej
fam_dead_df['procent'] = fam_dead_df.groupby('status')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
fam_dead_df

In [ ]:
# Przygotowanie danych dla wykresu pokazującego przeżywalność w zależności od statusu rodzinnego
fam_dead_pie_df = fam_dead_df.pivot(index='status', columns='survived', values='liczba')
fam_dead_pie_df

In [ ]:
# Generowanie wykresu pokazującego przeżywalność w zależności od statusu rodzinnego
fig, axes = plt.subplots(1, 2, figsize=(10, 3.5))
fig.suptitle('Przeżywalność w zależności od statusu rodzinnego', fontsize=14)
# Iterowanie przez wiersze df
for i, (index, row) in enumerate(fam_dead_pie_df.iterrows()):
    axes[i].pie(row, labels=['zginęli','przeżyli'], autopct='%1.1f%%', startangle=90)
    axes[i].set_title(f'pasażerowie {index}')
    axes[i].axis('equal')
# Dopasowanie wykresów, żeby się nie nachodziły
plt.tight_layout()
plt.show()

#### Wnioski
* z 794 pasażerów niespokrewnionych przeżyło 239, czyli 30,1%
* z 515 pasażerów spokrewnionych przeżyło 261, czyli 50,7%
* pasażerów spokrewnionych przeżyło znacznie więcej niż niespokrewnionych, prawdopodobnie z powodu większej obecności małych dzieci w tej grupie

### 5. Czy na przeżywalność dzieci miało wpływ posiadanie krewnych?

In [ ]:
# Utworzenie df zawierającego dane wyłącznie dzieci
child_df = df[df['age'] < 18]
# Zebranie danych dotyczących przeżywalności dzieci w zależności od statusu rodzinnego
ch_fam_surv_df = child_df.groupby([df['fam'].apply(lambda x: 'niespokrewnione' if x == '' else 'spokrewniene'), 'survived']).size().reset_index(name='liczba')
ch_fam_surv_df = ch_fam_surv_df.rename(columns={'fam' : 'status'})
# Dodanie kolumny z wyliczeniem wartości procentowej
ch_fam_surv_df['procent'] = ch_fam_surv_df.groupby('status')['liczba'].transform(lambda x: 100 * x / x.sum()).round(1)
ch_fam_surv_df

In [ ]:
# Przygotowanie danych dla wykresu pokazującego przeżywalność dzieci w zależności od statusu rodzinnego
ch_fam_surv_pie_df = ch_fam_surv_df.pivot(index='status', columns='survived', values='liczba')
ch_fam_surv_pie_df

In [ ]:
# Generowanie wykresu pokazującego przeżywalność dzieci w zależności od statusu rodzinnego
fig, axes = plt.subplots(1, 2, figsize=(10, 3.5))
fig.suptitle('Przeżywalność dzieci w zależności od statusu rodzinnego', fontsize=14)
# Iterowanie przez wiersze df
for i, (index, row) in enumerate(ch_fam_surv_pie_df.iterrows()):
    axes[i].pie(row, labels=['zginęły','przeżyły'], autopct='%1.1f%%', startangle=90)
    axes[i].set_title(f'dzieci {index}')
    axes[i].axis('equal')
# Dopasowanie wykresów, żeby się nie nachodziły
plt.tight_layout()
plt.show()

#### Wnioski
* z 32 dzieci niespokrewnionych przeżyło 13, czyli 40,6%
* ze 122 dzieci spokrewnionych przeżyło 68, czyli 55,7%
* status rodzinny dzieci miał nieduży wpływ na ich przeżywalność

### 7. Jaki był los służących i ich "rodzin"?

In [ ]:
# Utworzenie df jedynie z danymi służących i sprawdzenie ich liczby
maid_df = df[df['maid'] == 1].reset_index()
maid_df.shape[0]

In [ ]:
# Sprawdzenie ile służących przeżyło
(maid_df['survived'] == 1).sum()

In [ ]:
# Sprawdzenie procentowej przeżywalności służących
((maid_df['survived'] == 1).mean() * 100).round(1)

In [ ]:
# Utworzenie df członków rodzin i służących, którzy byli w tej samej łodzi ratunkowej (rodziny miały ten sam bilet, co służący)
df_grouped = df.groupby(['ticket', 'boat']).filter(lambda x: (x['maid'] == 1).any())
# Sprawdzenie ile było służących, którzy z członkami "swoich rodzin" trafili do tej samej łodzi ratunkowej
(df_grouped['maid'] == 1).sum()

In [ ]:
# Sprawdzenie rodziny Allison, z której uratował się jedynie niespełna roczny chłopiec dzięki jednej ze służących - pannie Cleaver
df[df['ticket'] == '113781']

#### Wnioski
* z grupy 23 służących przeżyło aż 20, czyli 87%
* 19 służących trafiło do łodzi ratunkowych razem z członkami "swoich rodzin"
* w rodzinie państwa Allison zginęli wszyscy oprócz niespełna rocznego synka uratowanego przez służącą - pannę Cleaver w łodzi nr 11

### 8. Czy miejsce w szalupie oznaczało ratunek?

In [ ]:
# Sprawdzenie, czy byli pasażerowie, którzy trafili do łodzi ratunkowej, ale nie przeżyli
df[(df['boat'].notna()) & (df['survived'] == 0)].reset_index()

#### Wnioski
* trafienie do szalupy nie gwarantowało przeżycia katastrofy
* 9 pasażerów, którzy trafili do różnych łodzi ratunkowych nie przeżyło katastrofy

## E. Wnioski końcowe
* z 1309 pasażerów katastrofę przeżyło 38,2% (500 osób) a zginęło 61,8% (809 osób)

### "Najpierw kobiety i dzieci" - ogólna zasada ewakuacji podczas katastrof
* w przypadku kobiet zasada ta zadziałała, ponieważ procentowo przeżyło ich prawie 4 razy więcej, niż mężczyzn
* w przypadku dzieci zasada ta nie zadziałała dobrze, ponieważ przeżyło ich 53% a dorosłych 39%
* jednak największa przeżywalność była w grupie wiekowej do 10 r.ż., bo aż 58%
  
### Pasażerowie w różnych klasach
##### Klasa biletu miała ogromny wpływ na losy pasażerów
* kobiety z klasy pierwszej i drugiej przeżyły prawie wszystkie, tymczasem z klasy trzeciej przeżyło nieco mniej niż połowa
* również dzieci z klasy pierwszej i drugiej przeżyły prawie wszystkie a z klasy trzeciej przeżyła jedynie 1/3
* podczas zajmowania miejsc w łodziach ratunkowych widać było wyraźny podział na klasy

### Bohaterska służąca
W rodzinie państwa Allison zginęli wszyscy oprócz niespełna rocznego chłopca uratowanego przez służącą - pannę Cleaver w łodzi nr 11
